python/ray/rllib/env/async_remote_env.py

```diff
@@ -0,0 +1,117 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import logging
+
+import ray
+from ray.rllib.env.base_env import BaseEnv, _DUMMY_AGENT_ID
+
+logger = logging.getLogger(__name__)
+
+
+class AsyncRemoteEnv(BaseEnv):
+    """Vector env that executes envs asynchronously in remote workers.
+
+    This provides dynamic batching of inference as observations are returned
+    from the remote simulator actors.
+    """
+
+    def __init__(self, make_env, num_envs, multiagent, sync):
+        self.make_local_env = make_env
+        if sync:
+            self.timeout = 9999999.0  # wait for all envs
+        else:
+            self.timeout = 0.0  # wait for only ready envs
+
+        def make_remote_env(i):
+            logger.info("Launching env {} in remote actor".format(i))
+            if multiagent:
+                return _RemoteMultiAgentEnv.remote(self.make_local_env, i)
+            else:
+                return _RemoteSingleAgentEnv.remote(self.make_local_env, i)
+
+        self.actors = [make_remote_env(i) for i in range(num_envs)]
+        self.pending = None  # lazy init
+
+    def poll(self):
+        if self.pending is None:
+            self.pending = {a.reset.remote(): a for a in self.actors}
+
+        # each keyed by env_id in [0, num_remote_envs)
+        obs, rewards, dones, infos = {}, {}, {}, {}
+        ready = []
+
+        # Wait for at least 1 env to be ready here
+        while not ready:
+            ready, _ = ray.wait(
+                list(self.pending),
+                num_returns=len(self.pending),
+                timeout=self.timeout)
+
+        # Get and return observations for each of the ready envs
+        env_ids = set()
+        for obj_id in ready:
+            actor = self.pending.pop(obj_id)
+            env_id = self.actors.index(actor)
+            env_ids.add(env_id)
+            ob, rew, done, info = ray.get(obj_id)
+            obs[env_id] = ob
+            rewards[env_id] = rew
+            dones[env_id] = done
+            infos[env_id] = info
+
+        logger.debug("Got obs batch for actors {}".format(env_ids))
+        return obs, rewards, dones, infos, {}
+
+    def send_actions(self, action_dict):
+        for env_id, actions in action_dict.items():
+            actor = self.actors[env_id]
+            obj_id = actor.step.remote(actions)
+            self.pending[obj_id] = actor
+
+    def try_reset(self, env_id):
+        obs, _, _, _ = ray.get(self.actors[env_id].reset.remote())
+        return obs
+
+
+@ray.remote(num_cpus=0)
```

Comment:
wouldn't be surprised if this requires GPU resources in the future (no need to address this rn)

Reply:
Yeah, I think so too.